### PR TITLE
ES CR: Reference properties

### PR DESCRIPTION
**Pull request by nezaniel**
This introduces references as first class citizens:
* Reference{PropertyName $name, PropertyCollection $properties, Node $node}
* ContentSubgraphInterface::findReferencingNodes and findReferencedNodes now return References instead of Nodes
* all entry points (FlowQuery, Change models,...) have been adjusted to extract the Nodes from the References for now

**Checklist**

- [ X ] Code follows the PSR-12 coding style
- [ X ] Tests have been created, run and adjusted as needed
- [ X ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php [hunk 1]
```diff
@@ -125,6 +125,7 @@ protected function getCurrentSubgraphs(): ContentSubgraphs
     {
         $currentSubgraphs = [];
         foreach ($this->getActiveContentGraphs() as $adapterName => $contentGraph) {
+            $contentGraph->disableCache();
```
Comment (skurfuerst) [inline Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php:128]: uh, why is this needed? I am not sure this is correct actually, but it might mask actual errors during the tests.

File: Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php [hunk 2]
```diff
@@ -669,4 +666,41 @@ protected function requireNodeAggregateToNotDisableDimensionSpacePoint(
...
     }
+
+    protected function validateReferenceProperties(
+        PropertyName $referenceName,
+        PropertyValuesToWrite $referenceProperties,
+        NodeTypeName $nodeTypeName
+    ): void {
+        $nodeType = $this->nodeTypeManager->getNodeType((string)$nodeTypeName);
+        // initialize node type
+        $nodeType->getOptions();
```
Comment (skurfuerst) [inline Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php:677]: do we still need this initialization?

File: Neos.ContentRepository/src/Feature/NodeReferencing/NodeReferencing.php [hunk 3]
```diff
@@ -83,34 +132,30 @@ public function handleSetNodeReferences(SetNodeReferences $command): CommandResu
...
-                    );
-                }
-
-                $domainEvents = DomainEvents::fromArray($events);
+
+                $event = DecoratedEvent::addIdentifier(
+                    new NodeReferencesWereSet(
+                        $command->contentStreamIdentifier,
+                        $command->sourceNodeAggregateIdentifier,
+                        $affectedOrigins,
```
Comment (skurfuerst) [inline Neos.ContentRepository/src/Feature/NodeReferencing/NodeReferencing.php:149]: ah, now I understand it. Still I guess we should add a comment in the event what the semantics of multiple affectedOrigins are.

File: Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php [hunk 4]
```diff
@@ -65,20 +79,43 @@ public function serializePropertyValues(
...
-                } else {
-                    // $propertyValue == null and not defined in NodeTypes -> we want to unset $propertyName!
-                    $serializedPropertyValues[$propertyName] = null;
-                }
+                // $propertyValue == null and defined in node types (we have a resolved type)
+                // -> we want to set the $propertyName to NULL
+                return new SerializedPropertyValue(null, (string)$propertyType);
+            }
+        } catch (PropertyTypeIsInvalid $exception) {
+            if ($exception->getCode() === 1630063406 && $propertyValue === null) {
```
Comment (skurfuerst) [inline Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php:92]: can we somehow change this code? Especially checking for an exception code is weird here IMHO

**Review by skurfuerst — CHANGES_REQUESTED** [review]
@nezaniel I love this feature and the change :) I have some nitpick comments / questions, which we can as well discuss via a call next week.

All the best,
Sebastian

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/GraphProjector.php [hunk 5]
```diff
@@ -579,32 +579,59 @@ public function whenNodePropertiesWereSet(NodePropertiesWereSet $event): void
...
-                $position++;
+
+                // set new
+                $position = 0;
+                foreach ($event->references as $reference) {
+                    $this->getDatabaseConnection()->insert('neos_contentgraph_referencerelation', [
+                        'name' => $event->referenceName,
+                        'position' => $position,
+                        'nodeanchorpoint' => $nodeAnchorPoint,
+                        'destinationnodeaggregateidentifier' => $reference->targetNodeAggregateIdentifier,
```
[the remark this answers is not in the document]
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/GraphProjector.php:628]: Let's unify it, should be consistent from command to read model imho

File: Neos.ContentRepository/src/Feature/NodeReferencing/Event/NodeReferencesWereSet.php [hunk 6]
```diff
@@ -27,30 +27,30 @@ final class NodeReferencesWereSet implements
     public function __construct(
         public readonly ContentStreamIdentifier $contentStreamIdentifier,
         public readonly NodeAggregateIdentifier $sourceNodeAggregateIdentifier,
-        public readonly OriginDimensionSpacePoint $sourceOriginDimensionSpacePoint,
-        public readonly NodeAggregateIdentifiers $destinationNodeAggregateIdentifiers,
+        public readonly OriginDimensionSpacePointSet $affectedSourceOriginDimensionSpacePoints,
```
[the remark this answers is not in the document]
Comment (nezaniel) [reply Neos.ContentRepository/src/Feature/NodeReferencing/Event/NodeReferencesWereSet.php:30]: That's for property scopes, a comment seems to be helpful

File: Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php:128]: Now that you mention it, there seem to be caching issues that should be resolved instead. Will have a look at it

File: Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php:677]: Nope, we don't :)

File: Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php (hunk 4 above)
Comment (nezaniel) [reply Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php:92]: I actually resurrected that part of code from a dead path...
Maybe we can simply remove the outer try/catch. 
The original idea from $back then is that by setting a non existent property to null we would be able to unset the property altogether. That didn't survive the introduction of property checks though and has zero test coverage. So... should we just drop it? Unsetting obsolete properties is covered by integrity checks I think

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/GraphProjector.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/GraphProjector.php:627]: Done, it's now source->target everywhere

File: Neos.ContentRepository/src/Feature/NodeReferencing/Event/NodeReferencesWereSet.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (nezaniel) [reply Neos.ContentRepository/src/Feature/NodeReferencing/Event/NodeReferencesWereSet.php:36]: Comment has been added

File: Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository/Tests/Behavior/Features/Bootstrap/CurrentSubgraphTrait.php]: Done

File: Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository/src/Feature/Common/ConstraintChecks.php]: Has been removed

File: Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php (hunk 4 above)
Comment (nezaniel) [reply Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php]: The "unset properties" feature has been restored on command level, tests will follow in a later PR

Comment (skurfuerst) [reply Neos.ContentRepository/src/Infrastructure/Property/PropertyConverter.php]: @nezaniel awesome :) Thanks!

**Review by skurfuerst — APPROVED** [review]
GREAT :) :) :)
